### PR TITLE
Arty GPIO & Freedom E SDK Improvements

### PR DESCRIPTION
**Pull request by alexdlukens**
**Related issue**: 

**Type of change**: other enhancement

**Impact**: other

**Release Notes**
 
Added GPIO functionality to Arty FPGA prototyping flow, used Freedom E SDK to verify functionality. Added script to enable multilib compiling flag & add additional env variables to RISCV GNU Toolchain build script.

### CODE REVIEW

File: fpga/src/main/scala/arty/Configs.scala
Comment (jerryz123) [inline fpga/src/main/scala/arty/Configs.scala:49]: Instead of adding this key which sets everything, lets add new classes
`class WithGPIOWidth(x: Int)`
and
`class WithGPIOIncludeIOF(iof: Boolean)` to the chipyard ConfigFragments.scala. These classes should read and override the base values set in `PeripheryGPIOKey`.

File: fpga/src/main/scala/arty/Configs.scala
Comment (jerryz123) [inline fpga/src/main/scala/arty/Configs.scala:77]: Lets avoid the `WithArtyGPIO` config above.

Add `WithArtyGPIOHarnessBinder` and `WithGPIOPassThrough` to `WithArtyTweaks`. 
Then in `ArtyWithGPIOConfig`, have
```
new chipyard.config.WithGPIOWidth(32) ++
new chipyard.config.WithGPIOIncludeIOF(true) ++
new TinyRocketArtyConfig
```

The purpose of this is to orthogonalize digital-level changes from IO-level changes from harness-level changes.

File: fpga/src/main/scala/arty/IOBinders.scala
Comment (jerryz123) [inline fpga/src/main/scala/arty/IOBinders.scala:60]: Can we add this to the chipyard/IOBinders.scala instead? Since this code is not specific to arty.
This should also be named `WithGPIOIOCells` since this code does instantiate IOCells in the ChipTop.

Also for generality let's do
``` 
system.iof.map ({ i =>
  i.iof_0.default()
  i.iof_1.default()
})
```

File: fpga/src/main/scala/arty/TestHarness.scala
Comment (jerryz123) [inline fpga/src/main/scala/arty/TestHarness.scala:9]: Is this necessary?

File: fpga/src/main/scala/arty/HarnessBinders.scala
Comment (jerryz123) [inline fpga/src/main/scala/arty/HarnessBinders.scala:108]: Suggestion: Add a require here to check that the underlying system was configured with GPIO width of 32.

**Review by jerryz123 — CHANGES_REQUESTED** [review]
Looks great overall. Some minor changes requested.

File: fpga/src/main/scala/arty/TestHarness.scala
Comment (alexdlukens) [reply fpga/src/main/scala/arty/TestHarness.scala:9]: This was unintended. Will quash changes to TestHarness.scala

File: fpga/src/main/scala/arty/Configs.scala
Comment (alexdlukens) [reply fpga/src/main/scala/arty/Configs.scala:77]: Moving `WithArtyGPIOHarnessBinder ` and `WithGPIOPassThrough` to WithArtyTweaks, without requiring the GPIO width to be 32 and IOFs to be included, will cause the TinyRocketArtyConfig to fail generation. I think the most reasonable approach would be to include the GPIO in TinyRocketArtyConfig and then remove the new "ArtyWithGPIOConfig".

**Comment by jerryz123** [comment]
This looks good to me, but I'm not sure what the status of the arty-spi-flash branch is @jamesdunn should decide what to do here.

**Comment by jamesdunn** [comment]
Hi all, very sorry for being late on this. This looks good to me as well. I think that it would be good to merge these changes into arty-spi-flash. The branch contains fixes to UART as well, so I think this GPIO enhancement is in line with this branch making IO fixes/improvements (UART, GPIO, SPI flash). I approve of merging these changes into arty-spi-flash, and once I resolve the arty-sim issues, I will simulate these IOs before submitting a merge request of both arty-spi-flash and arty-sim to dev.

**Comment by alexdlukens** [comment]
Thank you for the response, this all sounds great. I will move the GPIO into the WithArtyTweaks fragment when I get a chance to look at the code.

**Comment by joey0320** [comment]
Closing stale PR

**Pull request closed without merging**
